services/server/src/wss/handlers/session.ts

```diff
@@ -11,83 +11,116 @@ import { sessionId } from "@/validation/utils";
 
 const stdout = logger("wss");
 
-const onJoinSessionPayload = zod.object({ sessionId: sessionId });
-const onLeaveSessionPayload = zod.object({ sessionId: sessionId });
+const generalSessionPayload = zod.object({ sessionId: sessionId });
 
 export class Session extends WssHandler {
   public init(): Session {
+    this.socket.on(Wss.ClientEvent.PreJoinSession, this.onPreJoinSession.bind(this));
     this.socket.on(Wss.ClientEvent.JoinSession, this.onJoinSession.bind(this));
     this.socket.on(Wss.ClientEvent.LeaveSession, this.onLeaveSession.bind(this));
     return this;
   }
 
+  /*
+   *  This event listener shall be called whenever a user
+   *  renders the pre-join page in the frontend.
+   *  It joins the user socket to the session socket room.
+   */
+  async onPreJoinSession(data: unknown, callback?: Wss.AcknowledgeCallback) {
+    const result = await safe(async () => {
+      const { sessionId } = generalSessionPayload.parse(data);
+
+      const ack = await this.canJoinSession(sessionId);
+      if (ack.code !== Wss.AcknowledgeCode.Ok)
+        return this.call(callback, ack);
+
+      // join the user socket to the corresponding session socket.io room
+      this.socket.join(asSessionRoomId(sessionId));
+    });
+    if (result instanceof Error) stdout.error(result.message);
+  }
+
   /*
    *  This event listener will be called whenever a user
    *  joins a session. For instance, when
    *  users are joining a lesson, or a tutor is joining an
    *  interview.
    */
-  async onJoinSession(data: unknown) {
+  async onJoinSession(data: unknown, callback?: Wss.AcknowledgeCallback) {
     const result = await safe(async () => {
-      const { sessionId } = onJoinSessionPayload.parse(data);
+      const { sessionId } = generalSessionPayload.parse(data);
 
-      const user = this.user;
-      // todo: add ghost as a member of the session
-      if (!user) return stdout.error("user undefined!");
-      if (isGhost(user)) return stdout.warning("Unsupported");
+      const ack = await this.canJoinSession(sessionId);
+      if (ack.code !== Wss.AcknowledgeCode.Ok)
+        return this.call(callback, ack);
 
-      stdout.info(`User ${user.id} is joining session ${sessionId}.`);
+      // ensured to be self from the canJoinSession function
+      const user = this.user as IUser.Self;
```

Comment:
Let's avoid casting and perform the check one more time. 

```suggestion
      // ensured to be self from the canJoinSession function
      const user = this.user
      if (!isUser(user)) return 
```